### ISSUE TITLE
Iterative Inversion

### ISSUE THREAD

**Issue by aaa2222339** [betterze/TurboEdit#2]
Thank you very much for your work. I have some questions:
1、In Algorithm 1 Iterative Inversion , an equation is：
<img width="321" alt="截屏2024-08-28 17 15 02" src="https://github.com/user-attachments/assets/7193966f-4d75-4771-b5a3-afa521fc9d28">
But it is not introduced in the paper, how did this E come about?
2、What is the network structure of generator G  and Fsingle? The paper also did not provide a specific introduction

**Comment by betterze** [comment]
Thank you for your interest in our work.

1. Sorry, it is a typo. The E should be F, the inversion network.
2. The generator G is SDXL-turbo, we do not change its architecture. Please refer to its architecture in [SDXL](https://arxiv.org/abs/2307.01952) paper.
As we mentioned in section 4.1, the inversion network F is initiated from the generator G. It has the same architecture as G, and a few extra channels in the input and output layers (8 channels instead of 4 channels).

Sorry again for the typo.

**Comment by aaa2222339** [comment]
Thank you for your answer. I have another question: 
Since the generator G remains unchanged during the training process, is using the SDXL-turbo model and directly applying equation 3 in the paper equivalent to the generator G? If that's the case, why invent a generator G?

**Comment by betterze** [comment]
In the diffusion model, people usually consider the Unet as a denoising network, and the Unet is trained to predict the injected noise (epsilon prediction in equation 2). The epsilon hat theta is the SDXL-turbo.

We change this notion from noise prediction to clean sample prediction in the equation 3. The generator G is the same as the noise predictor with a simple change of notation. So G is SDXL-turbo changes to clean sample prediction.

The image generator G can perform generated image editing tasks. To edit a real image, we need to invert it into the generated trajectory of G. There is a lot of literature about real image inversion and editing in the stylegan ([StyleSpace](https://arxiv.org/abs/2011.12799), [PSP encoder](https://arxiv.org/abs/2008.00951)) and diffusion domain ([Null-text Inversion](https://null-text-inversion.github.io/), [ddpm](https://arxiv.org/abs/2304.06140))

**Comment by aaa2222339** [comment]
I encountered a problem while reproducing this paper. The loss function is formula 8, but during training, the weight of model F in the second step after optimizer. step() is nan. I also added clip_grad_norm_, but the problem has not been solved. What could be the problem? Or can you provide reference training code? Thanks !

**Comment by betterze** [comment]
I believe you trained the model with bfloat16, with high res and a large batch size. 

We use the reparameterization trick to regularize the noise and calculate the kl loss 

```
from diffusers.models.autoencoders.vae import DiagonalGaussianDistribution

moments= output_from_inversion_network
posterior=DiagonalGaussianDistribution(moments)
noise=posterior.sample()
kl_loss = posteriors.kl() # this step trends to overflow
```

pls refer the kl loss implementation [here](https://github.com/huggingface/diffusers/blob/v0.3.0/src/diffusers/models/vae.py#L354). It uses sum, when we have a lot of pixels, it will overflow. I rewrite the code to 

```
0.5 * torch.mean(
                    torch.pow(self.mean - other.mean, 2) / other.var
                    + self.var / other.var
                    - 1.0
                    - self.logvar
                    + other.logvar,
                    dim=[1, 2, 3],
                )
```

This solves the overflow problem for me. Please keep in mind that you need to change the kl_loss weight accordingly.

**Comment by aaa2222339** [comment]
Is this the correct method for calculating MSE loss? Because I used the MSE loss predicted by noise when training DM before（equation 2）, and I haven't written the MSE loss for x0, I don't know if this is correct。Or is there any MES loss code for x0 available for reference? This is my code：

with torch.no_grad():
        latents = vae.encode(image.to(accelerator.device, dtype=torch.float32)).latent_dist.sample()
        latents = latents * vae.config.scaling_factor
        x0 = latents.to(accelerator.device, dtype=weight_dtype) 
...
pred_original_sample = G(...)
mse_loss = torch.nn.functional.mse_loss(x0.float(), pred_original_sample.float(), reduction="mean")

**Comment by aaa2222339** [comment]
Thank you very much for your answer. Could you please share the training set you used? It seems that it is not explicitly stated in the paper. Additionally, during the use of Llava, hallucination output is often generated. How did you solve this problem?

**Comment by betterze** [comment]
1. We use a clean version of the Laion dataset, which removes the NSFW images. I believe you could directly use the Laion dataset for research purposes.

2. We also observed the Llava 1.5 captions are not very reliable, but it is the best caption model we could have one year ago. We use the raw Llava caption to train the inversion network without filtering. I suggest you try other vision language models for captioning, for example, Phi-3.
